File: app/src/main/java/com/animsh/moviem/adapters/CrewAdapter.java
package com.animsh.moviem.adapters;

import android.view.LayoutInflater;
import android.view.ViewGroup;

import androidx.annotation.NonNull;
import androidx.databinding.DataBindingUtil;
import androidx.recyclerview.widget.RecyclerView;

import com.animsh.moviem.R;
import com.animsh.moviem.databinding.ItemCrewContainerBinding;
import com.animsh.moviem.model.Crew;

import java.util.List;

/**
 * Created by animsh on 2/11/2021.
 */
public class CrewAdapter extends RecyclerView.Adapter<CrewAdapter.CrewViewHolder> {

    private List<Crew> crews;
    private LayoutInflater layoutInflater;

    public CrewAdapter(List<Crew> crews) {
        this.crews = crews;
    }


    @NonNull
    @Override
    public CrewViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        if (layoutInflater == null) {
            layoutInflater = LayoutInflater.from(parent.getContext());
        }
        ItemCrewContainerBinding crewBinding = DataBindingUtil.inflate(
                layoutInflater, R.layout.item_crew_container, parent, false
        );
        return new CrewViewHolder(crewBinding);
    }

    @Override
    public void onBindViewHolder(@NonNull CrewViewHolder holder, int position) {
        holder.bindCrew(crews.get(position));
    }

    @Override
    public int getItemCount() {
        return crews.size();
    }

    class CrewViewHolder extends RecyclerView.ViewHolder {

        private ItemCrewContainerBinding itemCrewContainerBinding;

        public CrewViewHolder(ItemCrewContainerBinding itemCrewContainerBinding) {
            super(itemCrewContainerBinding.getRoot());
            this.itemCrewContainerBinding = itemCrewContainerBinding;
        }

        public void bindCrew(Crew crew) {
            itemCrewContainerBinding.setCrew(crew);
            itemCrewContainerBinding.executePendingBindings();
        }
    }
}

File: app/src/main/java/com/animsh/moviem/adapters/CrewAdapter.kt
package com.animsh.moviem.adapters

import android.view.LayoutInflater
import android.view.ViewGroup
import androidx.recyclerview.widget.DiffUtil
import androidx.recyclerview.widget.RecyclerView
import com.animsh.moviem.databinding.ItemCrewContainerBinding
import com.animsh.moviem.models.movie.Crew
import com.animsh.moviem.util.MovieDiffUtil

/**
 * Created by animsh on 5/14/2021.
 */
class CrewAdapter : RecyclerView.Adapter<CrewAdapter.CrewViewHolder>() {

    private var crewList = emptyList<Crew>()

    class CrewViewHolder(private val binding: ItemCrewContainerBinding) :
        RecyclerView.ViewHolder(binding.root) {
        fun bindData(crew: Crew) {
            binding.crew = crew
            binding.executePendingBindings()
        }

        companion object {
            fun from(parent: ViewGroup): CrewViewHolder {
                val layoutInflater = LayoutInflater.from(parent.context)
                val binding = ItemCrewContainerBinding.inflate(layoutInflater, parent, false)
                return CrewViewHolder(binding)
            }
        }
    }

    override fun onCreateViewHolder(
        parent: ViewGroup,
        viewType: Int
    ): CrewViewHolder {
        return CrewViewHolder.from(parent)
    }

    override fun onBindViewHolder(holder: CrewViewHolder, position: Int) {
        val current = crewList[position]
        holder.bindData(current)
    }

    override fun getItemCount(): Int {
        return crewList.size
    }

    fun setData(newData: List<Crew>) {
        val diffUtil = MovieDiffUtil(crewList, newData)
        val diffUtilResult = DiffUtil.calculateDiff(diffUtil)
        crewList = newData
        diffUtilResult.dispatchUpdatesTo(this)
    }

    fun clearList() {
        crewList = emptyList()
    }
}